### PR TITLE
Using parallel utilities in GeoMechanicsApplication

### PR DESCRIPTION
**Pull request by Vahid-Galavi**
**📝 Description**
All parallel loops are replaced by parallel utilities. In addition, some refactoring has been done.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_processes/periodic_interface_process.hpp [hunk 1]
```diff
@@ -77,87 +77,67 @@ class PeriodicInterfaceProcess : public Process
     /// right after reading the model and the groups
     void ExecuteInitialize() override
     {
-        KRATOS_TRY;
+        KRATOS_TRY
 
-        int NCons = static_cast<int>(mr_model_part.Conditions().size());
-        ModelPart::ConditionsContainerType::iterator con_begin = mr_model_part.ConditionsBegin();
+        int nConditions = static_cast<int>(mrModelPart.Conditions().size());
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/periodic_interface_process.hpp:82]: seems unused

File: applications/GeoMechanicsApplication/custom_processes/periodic_interface_process.hpp [hunk 2]
```diff
@@ -77,87 +77,67 @@ class PeriodicInterfaceProcess : public Process
...
-            //                                 NodalStressMatrix(0,1)*NodalStressMatrix(0,1));
-            //     PrincipalStresses[1] = 0.5*(NodalStressMatrix(0,0)+NodalStressMatrix(1,1)) -
-            //                         sqrt(0.25*(NodalStressMatrix(0,0)-NodalStressMatrix(1,1))*(NodalStressMatrix(0,0)-NodalStressMatrix(1,1)) +
-            //                                 NodalStressMatrix(0,1)*NodalStressMatrix(0,1));
-            // }
-            // else
-            // {
+        KRATOS_TRY
+
+        int nConditions = static_cast<int>(mrModelPart.Conditions().size());
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/periodic_interface_process.hpp:111]: unused

File: applications/GeoMechanicsApplication/custom_processes/periodic_interface_process.hpp [hunk 3]
```diff
@@ -77,87 +77,67 @@ class PeriodicInterfaceProcess : public Process
...
 
-        #pragma omp parallel for
-        for (int i = 0; i < NElems; i++)
-        {
-            ModelPart::ElementsContainerType::iterator itElem = el_begin + i;
-            itElem->Set(ACTIVE,false);
+        if (nElements > 0) {
+            block_for_each(mrModelPart.Elements(), [&](Element& rElement) {
+                rElement.Set(ACTIVE,false);
+            });
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/periodic_interface_process.hpp:100]: I would use the VariableUtils here

File: applications/GeoMechanicsApplication/custom_processes/apply_write_result_scalar_process.hpp [hunk 4]
```diff
@@ -127,21 +121,16 @@ class ApplyWriteScalarProcess : public Process
...
+            const double Time = mrModelPart.GetProcessInfo()[TIME]/mTimeUnitConverter;
             ModelPart::NodesContainerType::iterator it_begin = mrModelPart.NodesBegin();
 
-            //#pragma omp parallel for
-            for (int i = 0; i<nNodes; i++)
-            {
+            for (int i = 0; i<nNodes; ++i) {
                 ModelPart::NodesContainerType::iterator it = it_begin + i;
 
                 double value = it->FastGetSolutionStepValue(var);
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/apply_write_result_scalar_process.hpp:136]: I would replace `std::endl` with `\n`. The difference is that the former one flushes the buffer which can be significantly slower

You could explicitly flush the file after the loop of you want to be sure

File: applications/GeoMechanicsApplication/custom_processes/apply_write_result_scalar_process.hpp [hunk 5]
```diff
@@ -127,21 +121,16 @@ class ApplyWriteScalarProcess : public Process
...
     void ExecuteFinalizeSolutionStep() override
     {
-        KRATOS_TRY;
-
-        const Variable<double> &var = KratosComponents< Variable<double> >::Get(mVariableName);
-
-        const double Time = mrModelPart.GetProcessInfo()[TIME]/mTimeUnitConverter;
+        KRATOS_TRY
 
         const int nNodes = static_cast<int>(mrModelPart.Nodes().size());
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/apply_write_result_scalar_process.hpp:126]: I don't think you need this cast

File: applications/GeoMechanicsApplication/custom_processes/apply_write_result_scalar_process.hpp [hunk 6]
```diff
@@ -83,33 +83,27 @@ class ApplyWriteScalarProcess : public Process
     /// right after reading the model and the groups
     void ExecuteInitialize() override
     {
-        KRATOS_TRY;
-
-        const Variable<double> &var = KratosComponents< Variable<double> >::Get(mVariableName);
-        const double Time = mrModelPart.GetProcessInfo()[TIME]/mTimeUnitConverter;
+        KRATOS_TRY
 
         const int nNodes = static_cast<int>(mrModelPart.Nodes().size());
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/apply_write_result_scalar_process.hpp:88]: samr

File: applications/GeoMechanicsApplication/custom_processes/apply_time_dependent_interpolate_line_pressure_process.hpp [hunk 7]
```diff
@@ -63,30 +63,21 @@ class ApplyTimeDependentInterpolateLinePressureProcess : public ApplyConstantInt
 
         const int nNodes = static_cast<int>(mrModelPart.Nodes().size());
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/custom_processes/apply_time_dependent_interpolate_line_pressure_process.hpp:64]: unused?

File: applications/GeoMechanicsApplication/documents/Kratos_Verification/Kratos_Verification.ilg [hunk 8]
```diff
@@ -1,6 +1,6 @@
-This is makeindex, version 2.15 [MiKTeX 2.9.6960 64-bit] (kpathsea + Thai support).
-Scanning style file C:/resources/latex/makeindex/deltares/deltares_nomentbl.ist...............
-** Input style error (file = C:/resources/latex/makeindex/deltares/deltares_nomentbl.ist, line = 51):
+This is makeindex, version 2.15 [MiKTeX 2.9.7140 64-bit] (kpathsea + Thai support).
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/documents/Kratos_Verification/Kratos_Verification.ilg:1]: can you please move this to the Documentation repo?

File: applications/GeoMechanicsApplication/documents/Kratos_Verification/Kratos_Verification.log [hunk 9]
```diff
@@ -1,16 +1,20 @@
-This is pdfTeX, Version 3.14159265-2.6-1.40.20 (MiKTeX 2.9.6960 64-bit) (preloaded format=pdflatex 2019.2.18)  27 AUG 2019 17:56
+This is pdfTeX, Version 3.14159265-2.6-1.40.20 (MiKTeX 2.9.7140 64-bit) (preloaded format=pdflatex 2019.8.22)  13 OCT 2021 10:02
```
Comment (philbucher) [inline applications/GeoMechanicsApplication/documents/Kratos_Verification/Kratos_Verification.log:1]: same

**Review by philbucher — APPROVED** [review]
:+1:

**Comment by Vahid-Galavi** [comment]
Thanks @philbucher for the review and nice comments. I will consider them in the next PR.

**Pull request merged**